src/hd.rs

```diff
@@ -181,111 +177,106 @@ impl HDWallet {
+                .with_context(|| format!("invalid descriptor {}", descriptor))?
+                .clone(),
             );
         }
-        for (xpub, rescan) in bare_xpubs {
+
+        // Xpubs
+        for (xyz, rescan) in xpubs {
+            // Change and receiving output descriptors
+            for change in 0..2 {
+                let descriptor = match xyz.script_type {
+                    ScriptType::P2pkh => {
+                        format!("pkh({}/{}/*)", xyz.extended_pubkey.to_string(), change)
+                    }
+                    ScriptType::P2wpkh => {
+                        format!("wpkh({}/{}/*)", xyz.extended_pubkey.to_string(), change)
+                    }
+                    ScriptType::P2shP2wpkh => {
+                        format!("sh(wpkh({}/{}/*))", xyz.extended_pubkey.to_string(), change)
+                    }
+                };
+                wallets.push(
+                    Self::from_descriptor(
+                        descriptor.clone(),
+                        network,
+                        gap_limit,
+                        initial_import_size,
+                        *rescan,
+                        rpc.clone(),
+                    )
+                    .with_context(|| format!("Invalid xpub-derived descriptor {}", descriptor))?
+                    .clone(),
+                );
+            }
+        }
+
+        // Bare xpubs
+        for (xyz, rescan) in bare_xpubs {
+            let descriptor = match xyz.script_type {
+                ScriptType::P2pkh => format!("pkh({})", xyz.extended_pubkey.to_string()),
+                ScriptType::P2wpkh => format!("wpkh({})", xyz.extended_pubkey.to_string()),
+                ScriptType::P2shP2wpkh => format!("sh(wpkh({}))", xyz.extended_pubkey.to_string()),
+            };
             wallets.push(
-                Self::from_bare_xpub(
-                    xpub.clone(),
+                Self::from_descriptor(
+                    descriptor.clone(),
                     network,
                     gap_limit,
                     initial_import_size,
                     *rescan,
+                    rpc.clone(),
                 )
-                .with_context(|| format!("invalid xpub {}", xpub))?,
+                .with_context(|| format!("Invalid bare-xpub-derived descriptor {}", descriptor))?
+                .clone(),
             );
         }
+
         if wallets.is_empty() {
-            error!("Please provide at least one xpub to track (via --xpub or --bare-xpub).");
-            bail!("no xpubs provided");
+            warn!("Please provide at least one descriptor to track (via --descriptors).");
+            bail!("No descriptors provided");
         }
         Ok(wallets)
     }
 
-    pub fn derive(&self, index: u32) -> ExtendedPubKey {
-        self.master
-            .derive_pub(&*EC, &[ChildNumber::from(index)])
-            .unwrap()
+    pub fn from_descriptor(
```

Comment:
How about calling this just `new`, since this is the primary way to initialize wallets with all other construction options delegating to it?

Reply:
Nice. HDWallet::new is currently implemented but not used. So this should just replace it. Fixed.